CDCgov/prime-simplereport
# Email notification for Patient Bulk Upload

# BACKEND PULL REQUEST

## Related Issue #4440 

Adds an email notification once patient bulk upload is complete. The email includes a link back to the application, where if logged in, the user can view the newly uploaded patients.

Update: Ready for review!

Also note that this is the same PR as #4758, I created this new branch to keep things tidy.

## Changes Proposed

-  adds an email template for patient bulk upload
- when save to the database save in the async method for patient bulk upload is complete, an email is sent to notify the user
- if processing fails, an email is sent to notify the user (screenshot [here](https://github.com/CDCgov/prime-simplereport/issues/4440#issuecomment-1324262939))

## Additional Information

Tests Added
- I added two unit tests for the new email template success + error
- Integration test to verify if processing and save is successful, success email gets triggered
- Integration test to verify if processing within the while loop fails, or if the saveAll to the db fails, error email gets triggered

Here is the sequence of events: before patient upload, after patient upload, and then the email that gets sent:
- before
![dev4 b4](https://user-images.githubusercontent.com/89797785/203332989-95ca9ecf-5622-4393-a452-1315b4571431.png)
- after
![dev4 after](https://user-images.githubusercontent.com/89797785/203333015-6dc14a18-7291-4295-838d-ef0702d06cc3.png)
- email
![dev4 email](https://user-images.githubusercontent.com/89797785/203333030-5d2f0a74-e159-40ec-ac12-88457bf8f336.png)

When an error occurs in processing, sending the error email is triggered. Note that error email cannot be triggered manually via the app since validation catches csv errors (forcing failure would require a code change to remove validation or force the db to fail). If you figure out a way to do this 🙌  let me know though. Since this isn't manually testable in an environment, here are the logs when I triggered errors in the code:
- processing error in while loop (turned off validation in `PatientBulkUploadService` and uploaded an invalid csv
![Screen Shot 2022-11-22 at 4 01 21 PM](https://user-images.githubusercontent.com/89797785/203425065-280e22fe-cd76-43c7-8f65-a815f4df5672.png)
- processing error in saveAll (forced an error in the `addPatientsAndPhoneNumbers` method)
![Screen Shot 2022-11-22 at 4 15 31 PM](https://user-images.githubusercontent.com/89797785/203425096-29174e35-830c-471c-846c-680be6f6480a.png)

## Testing

How should reviewers verify this PR?
- Go to **dev5**, upload patients via csv, and confirm that you receive an email notifying you when the upload is complete. Click the `See patients` link, which should take you directly to the patients page, if you are logged in to dev4.
- Note that error email cannot be triggered manually

## Checklist for Primary Reviewer
